File: composeApp/src/commonMain/kotlin/org/sesacteamproject/passmate/ui/payment/EarningsViewModel.kt
package org.sesacteamproject.passmate.ui.payment

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.sesacteamproject.passmate.auth.domain.usecase.IsSignedInUseCase
import org.sesacteamproject.passmate.core.model.onFailure
import org.sesacteamproject.passmate.core.model.onSuccess
import org.sesacteamproject.passmate.mvi.MviViewModel
import org.sesacteamproject.passmate.payment.domain.usecase.GetEarningsUseCase

class EarningsViewModel(
    private val getEarningsUseCase: GetEarningsUseCase,
    private val isSignedInUseCase: IsSignedInUseCase
) : MviViewModel<EarningsUiState, EarningsAction, EarningsEvent>(EarningsUiState()) {

    private var hasEntered = false

    private fun onEnter() {
        if (hasEntered) {
            return
        }
        hasEntered = true
        // 호스트(회원) 전용 가드 — 서버 검증이 최종 권위 (규칙 §8)
        if (!isSignedInUseCase.invoke()) {
            viewModelScope.launch {
                _event.emit(EarningsEvent.RequireSignIn)
            }
        } else {
            load()
        }
    }

    private fun load() {
        _uiState.update { it.copy(isLoading = true, loadFailed = false) }
        viewModelScope.launch {
            getEarningsUseCase.invoke(null)
                .onSuccess { earnings ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            loadFailed = false,
                            earnings = earnings,
                            items = earnings.items,
                            nextCursor = if (earnings.hasNext) earnings.nextCursor else null
                        )
                    }
                }
                .onFailure {
                    _uiState.update { it.copy(isLoading = false, loadFailed = true) }
                }
        }
    }

    private fun onLoadMore() {
        val state = _uiState.value
        val cursor = state.nextCursor

        if (cursor == null || state.isLoadingMore) {
            return
        }
        _uiState.update { it.copy(isLoadingMore = true) }
        viewModelScope.launch {
            getEarningsUseCase.invoke(cursor)
                .onSuccess { earnings ->
                    _uiState.update {
                        it.copy(
                            isLoadingMore = false,
                            items = it.items + earnings.items,
                            nextCursor = if (earnings.hasNext) earnings.nextCursor else null
                        )
                    }
                }
                .onFailure {
                    _uiState.update { it.copy(isLoadingMore = false) }
                    _event.emit(EarningsEvent.ShowNotice("내역을 더 불러오지 못했어요"))
                }
        }
    }

    private fun onClickViewAllHistory() {
        viewModelScope.launch {
            _event.emit(EarningsEvent.OpenCoinHistory)
        }
    }

    private fun onClickManageAccount() {
        viewModelScope.launch {
            _event.emit(EarningsEvent.OpenAccountSheet)
        }
    }

    private fun onAccountSaved() {
        load()
        viewModelScope.launch {
            _event.emit(EarningsEvent.ShowNotice("정산 계좌를 저장했어요"))
        }
    }

    private fun onNotice(message: String) {
        viewModelScope.launch {
            _event.emit(EarningsEvent.ShowNotice(message))
        }
    }

    override fun onAction(action: EarningsAction) {
        when (action) {
            is EarningsAction.Enter -> onEnter()
            is EarningsAction.Retry -> load()
            is EarningsAction.LoadMore -> onLoadMore()
            is EarningsAction.ClickViewAllHistory -> onClickViewAllHistory()
            is EarningsAction.ClickManageAccount -> onClickManageAccount()
            is EarningsAction.AccountSaved -> onAccountSaved()
            is EarningsAction.Notice -> onNotice(action.message)
        }
    }
}

File: iosApp/iosApp/ui/payment/EarningsViewModel.swift
import Combine
import Foundation
import Shared

// Compose EarningsViewModel.kt 미러 — 수익·정산 요약·내역 로드 (M-T4)
final class EarningsViewModel: ObservableObject {
    private let getEarningsUseCase: GetEarningsUseCase

    private let isSignedInUseCase: IsSignedInUseCase

    @Published private(set) var uiState = EarningsUiState()

    let event = PassthroughSubject<EarningsEvent, Never>()

    private var hasEntered = false

    private func onEnter() {
        if hasEntered {
            return
        }
        hasEntered = true
        // 호스트(회원) 전용 가드 — 서버 검증이 최종 권위 (규칙 §8)
        if isSignedInUseCase.invoke() {
            load()
        } else {
            event.send(.requireSignIn)
        }
    }

    private func load() {
        uiState.isLoading = true
        uiState.loadFailed = false
        getEarningsUseCase.invoke(cursor: nil) { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                let earnings = (result as? AppResultSuccess<AnyObject>)?.value as? Earnings

                if error == nil, let earnings {
                    self.uiState.isLoading = false
                    self.uiState.loadFailed = false
                    self.uiState.earnings = earnings
                    self.uiState.items = self.settlementItems(earnings)
                    self.uiState.nextCursor = earnings.hasNext ? earnings.nextCursor : nil
                } else {
                    self.uiState.isLoading = false
                    self.uiState.loadFailed = true
                }
            }
        }
    }

    private func onLoadMore() {
        guard let cursor = uiState.nextCursor, !uiState.isLoadingMore else { return }
        uiState.isLoadingMore = true
        getEarningsUseCase.invoke(cursor: cursor) { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self else { return }
                self.uiState.isLoadingMore = false
                if error == nil, let earnings = (result as? AppResultSuccess<AnyObject>)?.value as? Earnings {
                    self.uiState.items += self.settlementItems(earnings)
                    self.uiState.nextCursor = earnings.hasNext ? earnings.nextCursor : nil
                } else {
                    self.event.send(.showNotice(message: "내역을 더 불러오지 못했어요"))
                }
            }
        }
    }

    private func onAccountSaved() {
        load()
        event.send(.showNotice(message: "정산 계좌를 저장했어요"))
    }

    private func settlementItems(_ earnings: Earnings) -> [SettlementItem] {
        earnings.items.compactMap { $0 as? SettlementItem }
    }

    func action(_ action: EarningsAction) {
        switch action {
        case .enter:
            onEnter()
        case .retry:
            load()
        case .loadMore:
            onLoadMore()
        case .clickViewAllHistory:
            event.send(.openCoinHistory)
        case .clickManageAccount:
            event.send(.openAccountSheet)
        case .accountSaved:
            onAccountSaved()
        case let .notice(message):
            event.send(.showNotice(message: message))
        }
    }

    init(
        getEarningsUseCase: GetEarningsUseCase,
        isSignedInUseCase: IsSignedInUseCase
    ) {
        self.getEarningsUseCase = getEarningsUseCase
        self.isSignedInUseCase = isSignedInUseCase
    }
}
